pkg/authorization/authconfig.go

```diff
@@ -132,92 +133,103 @@ func (k *annotationAuthTypeDetector) Detect(_ context.Context, res *unstructured
 	return spi.Anonymous, nil
 }
 
-type expressionHostExtractor struct {
-	paths []string
-}
-
-func NewExpressionHostExtractor(paths []string) spi.HostExtractor {
-	return &expressionHostExtractor{paths: paths}
-}
+func UnifiedHostExtractor(extractors ...spi.HostExtractor) spi.HostExtractor { //nolint:gocognit //reason Inlined functions to avoid surface pollution
+	unique := func(in []string) []string {
+		store := map[string]bool{}
 
-func (k *expressionHostExtractor) Extract(target *unstructured.Unstructured) ([]string, error) {
-	hosts := []string{}
-
-	for _, path := range k.paths {
-		splitPath := strings.Split(path, ".")
-		extractedHosts, err := extractHosts(target, splitPath)
-
-		if err != nil {
-			return nil, fmt.Errorf("failed to extract hosts at path %s: %w", path, err)
+		for _, v := range in {
+			store[v] = true
 		}
 
-		parsedHosts, errAppend := appendHosts(hosts, extractedHosts...)
+		keys := make([]string, len(store))
 
-		if errAppend != nil {
-			return nil, fmt.Errorf("failed to append hosts %v: %w", extractedHosts, errAppend)
+		i := 0
+
+		for v := range store {
+			keys[i] = v
+			i++
 		}
 
-		hosts = parsedHosts
+		return keys
 	}
 
-	u := unique(hosts)
-	if len(u) == 0 {
-		return []string{"unknown.host.com"}, nil
+	isURI := func(host string) bool {
+		return strings.HasPrefix(host, "http://") || strings.HasPrefix(host, "https://")
 	}
 
-	return u, nil
-}
+	appendHosts := func(hosts []string, foundHosts ...string) ([]string, error) {
+		var errAllParse []error
 
-func extractHosts(target *unstructured.Unstructured, splitPath []string) ([]string, error) {
-	// extracting as string
-	if foundHost, found, err := unstructured.NestedString(target.Object, splitPath...); err == nil && found {
-		return []string{foundHost}, nil
-	}
+		for _, foundHost := range foundHosts {
+			if isURI(foundHost) {
+				parsedURL, errParse := url.Parse(foundHost)
+				if errParse != nil {
+					errAllParse = append(errAllParse, fmt.Errorf("failed to parse URL %s: %w", foundHost, errParse))
+				}
+
+				hosts = append(hosts, parsedURL.Host)
+			} else {
+				hosts = append(hosts, foundHost)
+			}
+		}
 
-	// extracting as slice of strings
-	if foundHosts, found, err := unstructured.NestedStringSlice(target.Object, splitPath...); err == nil && found {
-		return foundHosts, nil
+		return hosts, errors.Join(errAllParse...)
 	}
 
-	return nil, fmt.Errorf("neither string nor slice of strings found at path %v", splitPath)
-}
+	return func(target *unstructured.Unstructured) ([]string, error) {
+		var errAll []error
 
-func unique(in []string) []string {
-	store := map[string]bool{}
+		extractedHosts := []string{}
 
-	for _, v := range in {
-		store[v] = true
-	}
+		for _, extractor := range extractors {
+			extractedDelegateHosts, err := extractor(target)
```

Comment:
What do you mean by `Delegate` here?

Reply:
Delegate is wrong, but I couldn't think of a better name last night. Decorated maybe. 

UnifiedHostExtractor doesn't do any extraction it self, it's a wrapper to make a unique and unified view of multiple other HostExtractors. SubExtractor? RealExtractor? I don't know

Reply:
Or `extractedHosts` -> `combinedExtractedHosts` and then `extractedDelegateHosts` -> `extractedHosts`?

Reply:
A yes, easier way of looking at it

https://github.com/maistra/odh-platform/pull/49/commits/bd7965f4ce7558637e52e07733d317ed74ada0ea